Show progress bar while geometry cache is being loaded

Grandro opened the pull request:
These commits implement a progress bar while the geometry cache is being loaded.
The frontend sends an HTTP request in fixed time steps to the server in order to get the current progress.
If a request fails or the geometry cache was successfully loaded the frontend stops asking.

In order to process the new `/loadstatus` command a `GeomCache` has to be instantiated. Thus far the only implementation of instantiating a `GeomCache` `Server::loadCache` also started loading the data, so I split the method into `Server::createCache` and `Server::loadCache` to be able to only ensure a `GeomCache` exists when the loading status is being requested.

The loading bar will go from 0% to 100% twice because of the two stages caused by `GeomCache::parse` and `GeomCache::parseIds` which reset the progress. This info could also be displayed to the user.

patrickbr commented on src/qlever-petrimaps/GeomCache.cpp:
```diff
@@ -318,6 +317,14 @@ void GeomCache::parse(const char* c, size_t size) {
+  if (_totalSize == 0) {
+    return 0.0;
+  } else {
+    return (static_cast<double>(_curRow) / static_cast<double>(_totalSize) * 100);
```
I think it would be better if _curRow was now a `std::atomic<size_t>`, as concurrent read/writes can now happen.

patrickbr commented on src/qlever-petrimaps/server/Server.cpp:
```diff
@@ -1108,6 +1112,18 @@ util::http::Answer Server::handleExportReq(const Params& pars, int sock) const {
   return aw;
 }
 
+util::http::Answer Server::handleLoadStatusReq(const Params& pars) const {
```
When testing the app, I noticed that the percentage often "jumps back" to something like 60% shortly after it has reached 100%. I am not sure why this is happening, though.

patrickbr replied:
Ah, I just noticed that you already gave an explanation for this above :) The `parseId` stage is generally quite short, maybe we could cheat a bit and use the first 95% for the geometry progress, and the last 5% for the IDs.

Grandro replied:
If there is no value of telling the user in which stage of loading they currently are and there won't be any additional stages in the future I think that would be fine. Or maybe it is nice for debugging to let the user know in which stage they were if something goes wrong?

patrickbr commented on web/style.css:
```diff
@@ -120,12 +698,55 @@ h1, #msg-inner {
+    text-align: center;
+    font-size: 15px;
+    line-height: 30px;
+    color: white;
```
The color contrast is a bit bad here, maybe use `color:black`? You could also make the filled bar black, the empty bar white, and set the text color to `white` with `mix-blend-mode:difference`.

Grandro replied:
The second option you provided gives a nice effect, I am just not sure whether a black loading bar is odd to the user. I will keep trying around a little.

patrickbr commented on web/style.css:
```diff
@@ -120,12 +698,55 @@ h1, #msg-inner {
+    width: 0%;
+    height: 30px;
+    background-color: #04AA6D;
+}
```
We experimented with adding a `transition: width 500ms` here, this gives a nice effect.

patrickbr reviewed (commented):
Thanks for the PR, this works very well and already looks great. We noticed some minor things (see individual comments).

hannahbast commented:
Thanks a lot for this! When you merge this, please take care of two things:

1. Use "Squash and merge"
2. Write a meaningful commit message

Grandro commented:
@patrickbr I addressed the issues you mentioned in a new commit:
- Changed `_curRow` to a `std::atomic<size_t>`
- Merged the progress of both stages to a total progress (Default: Parse 95%, ParseIds 5%)
- Display the current stage to the user above the loading bar
- Changed the loading bar to `color:black`, text to `color:white` with `mix-blend-mode:difference`
- Added `transition: width 500ms` to loading bar

Are there still things that should be adjusted?
